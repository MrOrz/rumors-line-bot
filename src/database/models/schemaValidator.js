import Ajv from 'ajv';

const ajv = new Ajv();

const SCHEMAS = {
  userArticleLink: require('./userArticleLink.json'),
  userSettings: require('./userSettings.json'),
};

const ClassTable = {
  Date: Date,
};

ajv.addKeyword('instanceOf', {
  compile: function(className) {
    var Class = ClassTable[className];
    return function(data) {
      return data instanceof Class;
    };
  },
});

export function validate(schemaName, data) {
  const valid = ajv.validate(SCHEMAS[schemaName], data);
  return { valid, errors: ajv.errors };
}

export function compile(schemaName) {
  return data => {
    const validate = ajv.compile(SCHEMAS[schemaName]);
    const valid = validate(data);
    return { valid, errors: validate.errors };
  };
}
