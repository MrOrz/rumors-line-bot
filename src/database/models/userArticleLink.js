import Base from './base';

/**
 * https://g0v.hackmd.io/eIeU2g86Tfu5VnLazNfUvQ
 */
class UserArticleLink extends Base {
  /**
   *
   * @override
   */
  async create(data) {
    // Set default value
    const {
      createdAt = new Date(),
      lastViewedAt = null,
      lastRepliedAt = null,
      lastPositiveFeedbackRepliedAt = null,
      ...rest
    } = data;

    return super.create({
      ...rest,
      createdAt,
      lastViewedAt,
      lastRepliedAt,
      lastPositiveFeedbackRepliedAt,
    });
  }
}

export default new UserArticleLink('userArticleLink');
