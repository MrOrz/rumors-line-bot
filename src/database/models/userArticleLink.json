{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "cofacts/schemas/userArticleLink",
  "$comment": "design document https://g0v.hackmd.io/eIeU2g86Tfu5VnLazNfUvQ",
  "description": "A representation of an UserArticleLink",
  "title": "UserArticleLink",
  "type": "object",

  "properties": {
    "userId": {
      "type": "string"
    },
    "articleId": {
      "type": "string"
    },
    "createdAt": {
      "format": "date-time",
      "instanceOf": "Date"
    },
    "lastViewedAt": {
      "oneOf": [ {
        "type": "null"
      }, {
        "format": "date-time",
        "instanceOf": "Date"
      }]
    },
    "lastRepliedAt": {
      "oneOf": [ {
        "type": "null"
      }, {
        "format": "date-time",
        "instanceOf": "Date"
      }]
    },
    "lastPositiveFeedbackRepliedAt": {
      "oneOf": [ {
        "type": "null"
      }, {
        "format": "date-time",
        "instanceOf": "Date"
      }]
    }
  },
  "required": [
    "userId",
    "articleId",
    "createdAt",
    "lastViewedAt",
    "lastRepliedAt",
    "lastPositiveFeedbackRepliedAt"
  ]
}
