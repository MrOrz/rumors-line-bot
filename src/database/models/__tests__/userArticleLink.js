import Client from '../../mongoClient';
import MockDate from 'mockdate';
import { compile } from '../schemaValidator';
import userArticleLink from '../userArticleLink';

const userArticleLinkValidator = compile('userArticleLink');

describe('userArticleLink', () => {
  beforeAll(async () => {
    MockDate.set(612921600000);
  });

  afterAll(async () => {
    await (await Client.getInstance()).close();
  });

  it('[schema] should pass', async () => {
    const data = {
      userId: 'this_is_user_id',
      articleId: 'this_is_article_id',
      createdAt: new Date(),
      lastViewedAt: new Date(),
      lastRepliedAt: new Date(),
      lastPositiveFeedbackRepliedAt: new Date(),
    };
    const result = userArticleLinkValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[schema] should fail with wrong type of createdAt', async () => {
    const data = {
      userId: 'this_is_user_id',
      articleId: 'this_is_article_id',
      createdAt: '2020-01-01T20:10:18.314Z',
      lastViewedAt: new Date(),
      lastRepliedAt: new Date(),
      lastPositiveFeedbackRepliedAt: new Date(),
    };
    const result = userArticleLinkValidator(data);
    expect(result).toMatchSnapshot();
  });

  it('[model] should create a document', async () => {
    const result = await userArticleLink.create({
      userId: 'this_is_user_id',
      articleId: 'this_is_article_id',
    });
    delete result._id;
    expect(result).toMatchSnapshot();
  });

  it('[model] should fail to create a document', async () => {
    MockDate.set(612921600011);
    expect(
      userArticleLink.create({
        articleId: 'this_is_article_id',
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
