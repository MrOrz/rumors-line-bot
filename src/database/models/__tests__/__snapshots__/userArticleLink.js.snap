// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userArticleLink [model] createOrUpdateByUserIdAndArticleId() (upsert) 1`] = `
Object {
  "articleId": "articleId-3",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "lastViewedAt": 1989-06-04T00:01:00.000Z,
  "userId": "userId-3",
}
`;

exports[`userArticleLink [model] createOrUpdateByUserIdAndArticleId() 1`] = `
Object {
  "articleId": "articleId-2",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "lastViewedAt": 1989-06-04T00:01:00.000Z,
  "userId": "userId-2",
}
`;

exports[`userArticleLink [model] should create a document 1`] = `
Object {
  "articleId": "this_is_article_id",
  "createdAt": 1989-06-04T00:00:00.000Z,
  "lastViewedAt": 1989-06-04T00:00:00.000Z,
  "userId": "this_is_user_id",
}
`;

exports[`userArticleLink [model] should fail to create a document 1`] = `
"[
  {
    \\"keyword\\": \\"required\\",
    \\"dataPath\\": \\"\\",
    \\"schemaPath\\": \\"#/required\\",
    \\"params\\": {
      \\"missingProperty\\": \\"userId\\"
    },
    \\"message\\": \\"should have required property 'userId'\\"
  }
]"
`;

exports[`userArticleLink [schema] should fail with wrong type of createdAt 1`] = `
Object {
  "errors": Array [
    Object {
      "dataPath": ".createdAt",
      "keyword": "instanceOf",
      "message": "should pass \\"instanceOf\\" keyword validation",
      "params": Object {
        "keyword": "instanceOf",
      },
      "schemaPath": "#/properties/createdAt/instanceOf",
    },
  ],
  "valid": false,
}
`;

exports[`userArticleLink [schema] should pass 1`] = `
Object {
  "errors": null,
  "valid": true,
}
`;
